<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Cancer-dedicated gene set interpretation • oncoEnrichR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Cancer-dedicated gene set interpretation">
<meta name="description" content="oncoEnrichR performs annotation and prioritization of genesets discovered from high-throughput experiments within cancer biology. The tool queries a number of high-quality resources, including DepMap/Project Score (cellular fitness in genome-scale drop-out screens), Open Targets Platform (comprehensive target-drug associations, cancer hallmark evidence and target-tumor type associations), TCGA (genetic aberrations and gene co-expression patterns in human tumor samples), STRING/BioGRID (protein-protein interaction networks), curated protein complexes (CORUM/Compleat/ComplexPortal/hu.MAP2), COMPARTMENTS (subcellular compartments), Human Protein Atlas (prognostic gene expression associations), CellChatDB (ligand-receptor interactions), DoRothEA (regulatory interactions), synthetic lethality predictions, and Gene Ontology/MSigDB/REACTOME/KEGG/NetPath (functional enrichment/over-representation analysis with respect to gene function and cellular pathways). The package outputs both an interactive HTML report and an Excel workbook in which the user can interrogate the various types of annotations and analyses performed with the query geneset.">
<meta property="og:description" content="oncoEnrichR performs annotation and prioritization of genesets discovered from high-throughput experiments within cancer biology. The tool queries a number of high-quality resources, including DepMap/Project Score (cellular fitness in genome-scale drop-out screens), Open Targets Platform (comprehensive target-drug associations, cancer hallmark evidence and target-tumor type associations), TCGA (genetic aberrations and gene co-expression patterns in human tumor samples), STRING/BioGRID (protein-protein interaction networks), curated protein complexes (CORUM/Compleat/ComplexPortal/hu.MAP2), COMPARTMENTS (subcellular compartments), Human Protein Atlas (prognostic gene expression associations), CellChatDB (ligand-receptor interactions), DoRothEA (regulatory interactions), synthetic lethality predictions, and Gene Ontology/MSigDB/REACTOME/KEGG/NetPath (functional enrichment/over-representation analysis with respect to gene function and cellular pathways). The package outputs both an interactive HTML report and an Excel workbook in which the user can interrogate the various types of annotations and analyses performed with the query geneset.">
<meta property="og:image" content="https://sigven.github.io/oncoEnrichR/logo.png">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-N4MB4571K5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-N4MB4571K5');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">oncoEnrichR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html">Intro</a></li>
<li class="nav-item"><a class="nav-link" href="articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/input.html">Input</a></li>
    <li><a class="dropdown-item" href="articles/output.html">Output</a></li>
    <li><a class="dropdown-item" href="articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="articles/cancer_gene_rank.html">Cancer gene rank</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/oncoEnrichR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><br></p>
<div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="oncoenrichr-">oncoEnrichR 
<a class="anchor" aria-label="anchor" href="#oncoenrichr-"></a>
</h1>
</div>
<p><strong>oncoEnrichR</strong> is an R package for functional interrogation of human genesets in the context of cancer. It is primarily intended for exploratory analysis, interpretation, and prioritization of long gene lists, which represent a common output from many types of high-throughput cancer biology screens.</p>
<p><strong>oncoEnrichR</strong> can be used to interrogate results from e.g. genetic screens (siRNA/CRISPR), protein proximity labeling, or transcriptomics (differential expression). The tool queries a variety of high-quality data resources in order to assemble useful gene annotations and analyses in an interactive report (examples from the report shown below).</p>
<p>Web-based access to <strong>oncoEnrichR</strong> is available at <font size="4"> <a href="https://oncotools.elixir.no/tool_runner?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fsigven%2Foncoenrichr%2Foncoenrichr_wrapper%2F1.5.2" class="external-link"><strong>https://oncotools.elixir.no</strong></a></font></p>
<p><br></p>
<p><img src="reference/figures/oncoenrichr_overview.png" align="center" border="0"></p>
<p><br><br></p>
<div class="section level2">
<h2 id="questions-addressed-by-oncoenrichr">Questions addressed by oncoEnrichR<a class="anchor" aria-label="anchor" href="#questions-addressed-by-oncoenrichr"></a>
</h2>
<p>The contents of the analysis report provided by oncoEnrichR address the following scientific questions for a given gene list:</p>
<ul>
<li>Which diseases/tumor types are known to be associated with genes in the query set, and to what extent?</li>
<li>Which genes in the query set are known to have oncogenic or tumor suppressive roles, using evidence from the biomedical literature and curated resources?</li>
<li>Which genes in the query set are attributed with cancer hallmark evidence?</li>
<li>Which proteins in the query sets are druggable in different cancer conditions (early and late clinical development phases)? For other proteins in the query set, what is their likelihood of being druggable?</li>
<li>Which protein complexes involve proteins in the query set?</li>
<li>Which subcellular compartments (nucleus, cytosol, plasma membrane etc) are dominant localizations for proteins in the query set?</li>
<li>Which protein-protein interactions are known within the query set? Are there interactions between members of the query set and other cancer-relevant proteins (e.g. proto-oncogenes, tumor-suppressors or predicted cancer drivers)? Which proteins constitute hubs in the protein-protein interaction network?</li>
<li>Which known regulatory interactions (TF-target) are found within the query set, and what is their mode of regulation (repressive vs. stimulating)?</li>
<li>Are there occurrences of known ligand-receptor interactions within the query set?</li>
<li>Are there specific pathways, biological processes, or pre-defined molecular signatures that are enriched within the query set, as compared to a reference/background set?</li>
<li>Which members of the query set are frequently mutated in tumor sample cohorts (TCGA, SNVs/InDels, homozygous deletions, copy number amplifications)?</li>
<li>Which members of the query set are co-expressed (strong negative or positive correlations) with cancer-relevant genes (i.e. proto-oncogenes or tumor suppressors) in tumor sample cohorts (TCGA)?</li>
<li>Which members of the query set are associated with better/worse survival in different cancers, considering high or low gene expression levels, mutation, or copy number status in tumors?</li>
<li>Which members of the query set are predicted as partners of synthetic lethality interactions?</li>
<li>Which members of the query set are associated with cellular loss-of-fitness in CRISPR/Cas9 whole-genome drop out screens of cancer cell lines (i.e. reduction of cell viability elicited by a gene inactivation)? Which targets are prioritized therapeutic targets, considering fitness effects and genomic biomarkers in combination?</li>
</ul>
<p>See also the <a href="articles/output.html">output views</a> that addresses each of the questions above.</p>
</div>
<div class="section level2">
<h2 id="news">News<a class="anchor" aria-label="anchor" href="#news"></a>
</h2>
<ul>
<li>September 9th 2024: <a href="articles/CHANGELOG.html#version-1-5-2"><strong>1.5.2 release</strong></a>
</li>
<li>August 3rd 2024: <a href="articles/CHANGELOG.html#version-1-5-1"><strong>1.5.1 release</strong></a>
</li>
<li>August 16th 2023: <a href="articles/CHANGELOG.html#version-1-4-2"><strong>1.4.2 release</strong></a>
</li>
<li>June 18th 2023: <a href="articles/CHANGELOG.html#version-1-4-1"><strong>1.4.1 release</strong></a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use oncoEnrichR, please cite the following publication:</p>
<p>Sigve Nakken, Sveinung Gundersen, Fabian L. M. Bernal, Dimitris Polychronopoulos, Eivind Hovig, and Jørgen Wesche. <strong>Comprehensive interrogation of gene lists from genome-scale cancer screens with oncoEnrichR</strong> (2023). Int J Cancer. <a href="https://doi.org/10.1002/ijc.34666" class="external-link">doi:10.1002/ijc.34666</a></p>
</div>
<div class="section level2">
<h2 id="example-report">Example report<a class="anchor" aria-label="anchor" href="#example-report"></a>
</h2>
<p><a href="https://doi.org/10.5281/zenodo.13734871" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.13734871.svg" alt="DOI"></a></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<ul>
<li><a href="articles/installation.html">Installation instructions</a></li>
<li><a href="articles/running.html">How to run oncoEnrichR</a></li>
<li><a href="articles/output.html">Explore the various analysis outputs produced with oncoEnrichR</a></li>
<li><a href="articles/annotation_resources.html">Annotation resources available in oncoEnrichR</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>sigven AT ifi.uio.no</p>
</div>
<div class="section level2">
<h2 id="funding-and-collaboration">Funding and Collaboration<a class="anchor" aria-label="anchor" href="#funding-and-collaboration"></a>
</h2>
<p>oncoEnrichR is supported by the <a href="https://www.med.uio.no/cancell/english/" class="external-link">Centre for Cancer Cell Reprogramming</a> at the <a href="https://www.uio.no" class="external-link">University of Oslo</a>/<a href="https://radium.no" class="external-link">Oslo University Hospital</a>, and <a href="https://elixir.no/organization/organisation/elixir-uio" class="external-link">Elixir Norway (Oslo node)</a>.</p>
<p><br><br></p>
<p float="left">
</p>
<p><a href="https://www.med.uio.no/cancell/english/" class="external-link"> <img src="img/can-cell.png" width="150"></a>       <a href="https://elixir.no/organization/organisation/elixir-uio" class="external-link"> <img src="img/elixir_norway.png" width="200"></a></p>

</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sigven/oncoEnrichR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/sigven/oncoEnrichR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><p><a href="authors.html#citation">Citing oncoEnrichR</a></p></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/sigven" class="external-link">Sigve Nakken</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8468-2050" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
