<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Function that interrogates a list of human gene identifiers for
cancer relevance. Multiple perspectives are offered, including
tumor aberration and co-expression patterns, druggability,
protein-protein interactions, gene fitness effects, regulatory
interactions, subcellular compartment enrichment, pathway enrichment,
synthetic lethality interactions, prognostic associations, and more."><title>Interrogate a gene list for cancer relevance — onco_enrich • oncoEnrichR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Interrogate a gene list for cancer relevance — onco_enrich"><meta property="og:description" content="Function that interrogates a list of human gene identifiers for
cancer relevance. Multiple perspectives are offered, including
tumor aberration and co-expression patterns, druggability,
protein-protein interactions, gene fitness effects, regulatory
interactions, subcellular compartment enrichment, pathway enrichment,
synthetic lethality interactions, prognostic associations, and more."><meta property="og:image" content="https://sigven.github.io/oncoEnrichR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script async src="https://www.googletagmanager.com/gtag/js?id=G-N4MB4571K5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-N4MB4571K5');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">oncoEnrichR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/running.html">Running</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/input.html">Input</a>
    <a class="dropdown-item" href="../articles/output.html">Output</a>
    <a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a>
    <a class="dropdown-item" href="../articles/cancer_gene_rank.html">Cancer gene rank</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sigven/oncoEnrichR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Interrogate a gene list for cancer relevance</h1>
      <small class="dont-index">Source: <a href="https://github.com/sigven/oncoEnrichR/blob/HEAD/R/onco_enrichr.R" class="external-link"><code>R/onco_enrichr.R</code></a></small>
      <div class="d-none name"><code>onco_enrich.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Function that interrogates a list of human gene identifiers for
cancer relevance. Multiple perspectives are offered, including
tumor aberration and co-expression patterns, druggability,
protein-protein interactions, gene fitness effects, regulatory
interactions, subcellular compartment enrichment, pathway enrichment,
synthetic lethality interactions, prognostic associations, and more.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">onco_enrich</span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  oeDB <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_id_type <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  ignore_id_err <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  html_floating_toc <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  html_report_theme <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  project_title <span class="op">=</span> <span class="st">"_Project title_"</span>,</span>
<span>  project_owner <span class="op">=</span> <span class="st">"_Project owner_"</span>,</span>
<span>  project_description <span class="op">=</span> <span class="st">"_Project description_"</span>,</span>
<span>  bgset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bgset_id_type <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  bgset_description <span class="op">=</span> <span class="st">"All protein-coding genes"</span>,</span>
<span>  enrichment_p_value_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  enrichment_p_value_adj <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  enrichment_q_value_cutoff <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  enrichment_min_geneset_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  enrichment_max_geneset_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  enrichment_plot_num_terms <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  enrichment_simplify_go <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  subcellcomp_min_confidence <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  subcellcomp_min_channels <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  subcellcomp_show_cytosol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  regulatory_min_confidence <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>  fitness_max_score <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  ppi_add_nodes <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  ppi_string_min_score <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  ppi_string_network_type <span class="op">=</span> <span class="st">"functional"</span>,</span>
<span>  ppi_biogrid_min_evidence <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  ppi_node_shadow <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ppi_show_drugs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ppi_show_isolated_nodes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_ppi <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_disease <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_top_diseases_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_cancer_hallmarks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_drug <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_enrichment <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_aberration <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_coexpression <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_cell_tissue <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_ligand_receptor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_regulatory <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_unknown_function <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_prognostic <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_subcell_comp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_synleth <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_fitness <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_complex <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_domain <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>query</dt>
<dd><p>character vector with gene/query identifiers (minimum 2, maximum 1000)</p></dd>


<dt>oeDB</dt>
<dd><p>oncoEnrichR data repository object - as returned from <code><a href="load_db.html">load_db()</a></code></p></dd>


<dt>query_id_type</dt>
<dd><p>character indicating source of query (one of
"uniprot_acc", "symbol","entrezgene", or "ensembl_gene", "ensembl_mrna",
"refseq_mrna", "ensembl_protein", "refseq_protein")</p></dd>


<dt>ignore_id_err</dt>
<dd><p>logical indicating if analysis should
continue when uknown query identifiers are encountered</p></dd>


<dt>html_floating_toc</dt>
<dd><p>logical - float the table of contents to the left
of the main document content (HTML report). The floating table of contents
will always be visible even when the document is scrolled</p></dd>


<dt>html_report_theme</dt>
<dd><p>Bootswatch theme for HTML report (any of
"bootstrap", "cerulean", "cosmo", "default",
"flatly", "journal", "lumen", "paper", "sandstone", "simplex",
"spacelab", "united", "yeti")</p></dd>


<dt>project_title</dt>
<dd><p>project title (title of report)</p></dd>


<dt>project_owner</dt>
<dd><p>name of project owner</p></dd>


<dt>project_description</dt>
<dd><p>project background information</p></dd>


<dt>bgset</dt>
<dd><p>character vector with gene identifiers, used as
reference/background for enrichment/over-representation analysis</p></dd>


<dt>bgset_id_type</dt>
<dd><p>character indicating source of background
("uniprot_acc", "symbol", "entrezgene",
"ensembl_gene", "ensembl_mrna", "refseq_mrna", "ensembl_protein",
"refseq_protein"), default: "symbol"</p></dd>


<dt>bgset_description</dt>
<dd><p>character indicating type of background
(e.g. "All lipid-binding proteins (n = 200)")</p></dd>


<dt>enrichment_p_value_cutoff</dt>
<dd><p>cutoff p-value for enrichment/
over-representation analysis (default: 0.05)</p></dd>


<dt>enrichment_p_value_adj</dt>
<dd><p>one of "holm", "hochberg", "hommel",
"bonferroni", "BH", "BY", "fdr", "none" (clusterProfiler, default: "BH")</p></dd>


<dt>enrichment_q_value_cutoff</dt>
<dd><p>cutoff q-value for enrichment
analysis (clusterProfiler, default: 0.2)</p></dd>


<dt>enrichment_min_geneset_size</dt>
<dd><p>minimal size of geneset annotated by term
for testing in enrichment/over-representation analysis (clusterProfiler,
default: 10)</p></dd>


<dt>enrichment_max_geneset_size</dt>
<dd><p>maximal size of geneset annotated by term
for testing in enrichment/over-representation analysis (clusterProfiler,
default: 500)</p></dd>


<dt>enrichment_plot_num_terms</dt>
<dd><p>number of top enriched Gene Ontology terms
(max) to show in enrichment barplot (default: 15)</p></dd>


<dt>enrichment_simplify_go</dt>
<dd><p>remove highly similar GO terms in results from
GO enrichment/over-representation analysis (default: TRUE)</p></dd>


<dt>subcellcomp_min_confidence</dt>
<dd><p>minimum confidence level for subcellular
compartment annotation in COMPARTMENTS (min = 3, max = 5, default: 3)</p></dd>


<dt>subcellcomp_min_channels</dt>
<dd><p>minimum number of channels that support a
subcellular compartment annotation in COMPARTMENTS (min = 1,
max = 3, default: 1)</p></dd>


<dt>subcellcomp_show_cytosol</dt>
<dd><p>logical indicating if subcellular heatmap
should show highlight proteins located in the cytosol or not (default: FALSE)</p></dd>


<dt>regulatory_min_confidence</dt>
<dd><p>minimum confidence level for regulatory
interactions (TF-target) retrieved from DoRothEA ('A','B','C', or 'D',
default: 'D')</p></dd>


<dt>fitness_max_score</dt>
<dd><p>maximum loss-of-fitness score (scaled Bayes factor
from BAGEL) for genes retrieved from DepMap/Project Score, default:-2</p></dd>


<dt>ppi_add_nodes</dt>
<dd><p>number of nodes to add to target set when computing the
protein-protein interaction network (STRING/BioGRID, default: 30)</p></dd>


<dt>ppi_string_min_score</dt>
<dd><p>minimum score (between 0 and 1) for confidence of
retrieved protein-protein interactions (STRING, default: 0.9)</p></dd>


<dt>ppi_string_network_type</dt>
<dd><p>type of network to show for interactions in
STRING ('functional' or 'physical', default: 'functional')</p></dd>


<dt>ppi_biogrid_min_evidence</dt>
<dd><p>minimum number of evidence items required
for protein-protein interactions retrieved (BioGRID, default: 3)</p></dd>


<dt>ppi_node_shadow</dt>
<dd><p>show shadow for nodes in the displayed PPI
network (default: TRUE)</p></dd>


<dt>ppi_show_drugs</dt>
<dd><p>logical indicating if targeted drugs (&gt;= phase 3)
should be displayed in protein-protein interaction networks (default: TRUE)</p></dd>


<dt>ppi_show_isolated_nodes</dt>
<dd><p>logical indicating if targets/nodes without
any interactions should be displayed in the protein-protein
interaction networks (default: FALSE)</p></dd>


<dt>show_ppi</dt>
<dd><p>logical indicating if report should contain protein-protein
interaction views of the query set (STRING and BioGRID, default: TRUE)</p></dd>


<dt>show_disease</dt>
<dd><p>logical indicating if report should contain disease
associations (Open Targets Platform, association_score &gt;= 0.05, support
from at least two data types), and tumor suppressor/oncogene annotations (
default: TRUE)</p></dd>


<dt>show_top_diseases_only</dt>
<dd><p>logical indicating if report should contain
top (n = 20) disease associations only pr. query gene
default: TRUE)</p></dd>


<dt>show_cancer_hallmarks</dt>
<dd><p>logical indicating if report should
contain annotations/evidence of cancer hallmarks per query gene
(COSMIC/Open Targets Platform, default: TRUE</p></dd>


<dt>show_drug</dt>
<dd><p>logical indicating if report should contain targeted
cancer drug information (default: TRUE)</p></dd>


<dt>show_enrichment</dt>
<dd><p>logical indicating if report should contain functional
enrichment/over-representation analysis (MSigDB, GO, KEGG, REACTOME,
NetPath, WikiPathways, default: TRUE)</p></dd>


<dt>show_aberration</dt>
<dd><p>logical indicating if report should contain TCGA
aberration plots (amplifications/deletions, default: TRUE)</p></dd>


<dt>show_coexpression</dt>
<dd><p>logical indicating if report should contain TCGA
co-expression data (RNAseq) of query set with oncogenes/tumor
suppressor genes (default: TRUE)</p></dd>


<dt>show_cell_tissue</dt>
<dd><p>logical indicating if report should contain
tissue-specificity and single cell-type specificity assessments
(Human Protein Atlas) of target genes (default: FALSE)</p></dd>


<dt>show_ligand_receptor</dt>
<dd><p>logical indicating if report should contain
ligand-receptor interactions (CellChatDB, default: TRUE)</p></dd>


<dt>show_regulatory</dt>
<dd><p>logical indicating if report should contain data on
transcription factor (TF) - target interactions relevant for the
query set (DoRothEA, default: TRUE)</p></dd>


<dt>show_unknown_function</dt>
<dd><p>logical indicating if report should highlight
target genes with unknown or poorly defined functions
(GO/Uniprot KB/NCBI, default: TRUE)</p></dd>


<dt>show_prognostic</dt>
<dd><p>logical indicating if mRNA-based (single-gene)
prognostic associations to cancer types should be listed
(Human Protein Atlas/TCGA, default: TRUE</p></dd>


<dt>show_subcell_comp</dt>
<dd><p>logical indicating if report should provide
subcellular compartment annotations (COMPARTMENTS, default: TRUE)</p></dd>


<dt>show_synleth</dt>
<dd><p>logical indicating if report should list overlap with
predicted synthetic lethality interactions (gene paralogs only,
De Kegel et al., Cell Systems, 2021). Default: TRUE</p></dd>


<dt>show_fitness</dt>
<dd><p>logical indicating if report should provide fitness
scores and target priority scores from CRISPR/Cas9 loss-of-fitness
screens (DepMap/Project Score, default: TRUE)</p></dd>


<dt>show_complex</dt>
<dd><p>logical indicating if report should provide target
memberships in known protein complexes
(ComplexPortal/Compleat/hu.MAP2/PDB/CORUM, default: TRUE)</p></dd>


<dt>show_domain</dt>
<dd><p>logical indicating if report should provide target
memberships in known protein domains (Pfam, default: TRUE)</p></dd>


<dt>...</dt>
<dd><p>arguments for Galaxy/web-based processing</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An oncoEnrichR report list object, with two main elements,
<code>data</code> and <code>config</code>. <code>data</code> contains data that goes into
each section of the output reports, <code>config</code> contains all
metadata for annotation resources used.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

