### Tumor aberration frequencies

* The somatic mutation frequencies of the target genes in ~10,000 tumor samples have been calculated with data from [The Cancer Genome Atlas (TCGA)](https://portal.gdc.cancer.gov/)

<br><br>

#### SNVs/InDels - oncoplots {.tabset}

* Frequency of somatic SNVs/InDels in the query set genes (top mutated) are illustrated with [oncoplots](https://www.bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/oncoplots.html)
* Query set mutation frequencies are sorted by type of diagnosis (i.e. cancer subtypes)
* The frequency of transitions/transversions is also shown per sample

<br><br>

```{r tcga_oncoplot_setup, echo = F, eval = T}

tcga_oncoplot_data <- onc_enrich_report[["data"]][["tcga"]][["aberration"]][["table"]][["snv_indel"]]
maf_url <- paste0("https://github.com/sigven/oncoEnrichR/raw/",
                  onc_enrich_report[["config"]][["resources"]][["oncoEnrichR"]][["version"]],
                  "/data-raw/maf/")

## oncoplot arguments
gene_mar <- 10
annotationFontSize <- 2.5
sepwd_genes <- 0.7
fontSize <- 1.5
legendFontSize <- 2.4
legend_height <- 5
anno_height <- 0.6

fheight <- 26
fwidth <- 18

for(n in names(tcga_oncoplot_data)){
  tcga_oncoplot_data[[n]]$fheight <- fheight
  n_genes <- NROW(tcga_oncoplot_data[[n]][['top_mutated_genes']])
  if(n_genes < 40){
    tcga_oncoplot_data[[n]]$fheight <- 24
  }
  if(n_genes < 30){
    tcga_oncoplot_data[[n]]$fheight <- 21
  }
  if(n_genes < 20){
    tcga_oncoplot_data[[n]]$fheight <- 17
  }
  if(n_genes < 10){
    tcga_oncoplot_data[[n]]$fheight <- 14
  }
  
  
  
}



```


##### Breast

```{r tcga_oncoplot_breast, echo = F, eval = NROW(tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Breast"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Breast"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F, 
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_breast_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Colon/Rectum

```{r tcga_oncoplot_colon, echo = F, eval = NROW(tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Colon/Rectum"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Colon/Rectum"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_colon_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Lung

```{r tcga_oncoplot_lung, echo = F, eval = NROW(tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Lung"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Lung"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_lung_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Skin

```{r tcga_oncoplot_skin, echo = F, eval = NROW(tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Skin"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Skin"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_skin_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Esophagus/Stomach

```{r tcga_oncoplot_stomach, echo = F, eval = NROW(tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Esophagus/Stomach"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Esophagus/Stomach"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_esophagus_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Cervix

```{r tcga_oncoplot_cervix, echo = F, eval = NROW(tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Cervix"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Cervix"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   draw_titv = T, bgCol = "gray93",
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_cervix_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Prostate

```{r tcga_oncoplot_prostate, echo = F, eval = NROW(tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Prostate"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Prostate"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_prostate_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Ovary/Fallopian Tube

```{r tcga_oncoplot_ovary, echo = F, eval = NROW(tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Ovary/Fallopian Tube"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_ovary_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Uterus


```{r tcga_oncoplot_uterus, echo = F, eval = NROW(tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Uterus"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Uterus"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_uterus_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Pancreas


```{r tcga_oncoplot_pancreas, echo = F, eval = NROW(tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Pancreas"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Pancreas"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_pancreas_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Soft Tissue


```{r tcga_oncoplot_softtissue, echo = F, eval = NROW(tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Soft Tissue"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Soft Tissue"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_softtissue_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Myeloid


```{r tcga_oncoplot_myeloid, echo = F, eval = NROW(tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Myeloid"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Myeloid"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_myeloid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### CNS/Brain


```{r tcga_oncoplot_cns, echo = F, eval = NROW(tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["CNS/Brain"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["CNS/Brain"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_brain_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Liver


```{r tcga_oncoplot_liver, echo = F, eval = NROW(tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Liver"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Liver"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   draw_titv = T, bgCol = "gray93",
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_liver_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Kidney


```{r tcga_oncoplot_kidney, echo = F, eval = NROW(tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Kidney"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Kidney"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_kidney_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Lymphoid


```{r tcga_oncoplot_lymphoid, echo = F, eval = NROW(tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Lymphoid"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Lymphoid"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_lymphoid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Head and Neck


```{r tcga_oncoplot_headneck, echo = F, eval = NROW(tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Head and Neck"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Head and Neck"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```


```{r tcga_headneck_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Biliary Tract


```{r tcga_oncoplot_biliarytract, echo = F, eval = NROW(tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Biliary Tract"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Biliary Tract"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   draw_titv = T, bgCol = "gray93",
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_biliary_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Bladder/Urinary Tract


```{r tcga_oncoplot_bladder, echo = F, eval = NROW(tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Bladder/Urinary Tract"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Bladder/Urinary Tract"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   draw_titv = T, bgCol = "gray93",
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_bladder_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Pleura


```{r tcga_pleura, echo = F, eval = NROW(tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Pleura"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Pleura"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   draw_titv = T, bgCol = "gray93",
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_pleura_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Thyroid


```{r tcga_thyroid, echo = F, eval = NROW(tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]) >= 5, fig.height = tcga_oncoplot_data[["Thyroid"]]$fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Thyroid"]][['code']]
maf <- readRDS(url(paste0(maf_url, code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]$symbol,
                   drawRowBar = F, drawColBar = F,
                   sortByAnnotation = T,
                   draw_titv = T, bgCol = "gray93",
                   clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_thyroid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]) < 5}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


#### SNVs/InDels - recurrent variants

<br>

* Browse recurrent, protein-coding somatic SNVs/InDels from TCGA in the query set 
  * Variants are listed as one record _per tissue/site_, in effect making the same variant occurring in multiple rows
  * Only variants with a site-specific frequency >= 2 are shown
  * Variants can be filtered based on various properties, e.g. __Site/tissue variant recurrence__, as well as overall recurrence across all tumor sites (column __Pancancer variant recurrence__)
  * Notably, each variant have been annotated/classified with a 
    1. _loss-of-function_ status, based on the [LOFTEE plugin in VEP](https://github.com/konradjk/loftee)
    2. Status as <span style="background-color:black; padding:2x; color:white">&nbsp;<b>somatic mutation hotspots&nbsp;</b></span> in cancer, according to [cancerhotspots.org](https://cancerhotspots.org). Format: _\<GENE\>|\<CODON\>|\<Q-VALUE\>_
* Top 2,500 recurrent variants are listed here (all variants are listed in the Excel output of *oncoEnrichR*)
    
<br><br>


```{r filter_tcga_recurrence, echo=F, results='asis'}

missing_recurrent_tcga_variants <- TRUE
if(nrow(onc_enrich_report[['data']][['tcga']][['recurrent_variants']]) > 0){
  missing_recurrent_tcga_variants <- FALSE
}

```

```{r table_browse_recurrent_vars, echo=F, results = "asis", eval = NROW(onc_enrich_report[['data']][['tcga']][['recurrent_variants']]) > 0}
library(crosstalk)


vars <- onc_enrich_report[['data']][['tcga']][['recurrent_variants']] %>%
  dplyr::mutate(SITE_RECURRENCE = as.integer(SITE_RECURRENCE))

mutation_hotspot_levels <- levels(as.factor(
  vars$MUTATION_HOTSPOT))


vars <- vars %>%
  dplyr::arrange(desc(SITE_RECURRENCE), desc(LOSS_OF_FUNCTION)) %>%
  dplyr::mutate(CONSEQUENCE = stringr::str_replace_all(
    CONSEQUENCE,"&",", "
  )) %>%
  head(2500)

if(length(mutation_hotspot_levels) > 0){
  vars <- vars %>%
    dplyr::arrange(MUTATION_HOTSPOT, desc(SITE_RECURRENCE))
}

recurrent_vars_tcga <- crosstalk::SharedData$new(
  vars)

crosstalk::bscols(
  list(
    crosstalk::filter_select("SYMBOL","Gene symbol",recurrent_vars_tcga, ~SYMBOL),
    crosstalk::filter_select("CONSEQUENCE", "Variant consequence", recurrent_vars_tcga, ~CONSEQUENCE),
    crosstalk::filter_select("PRIMARY_SITE", "Primary site/tissue", recurrent_vars_tcga, ~PRIMARY_SITE),
    crosstalk::filter_select("MUTATION_HOTSPOT", 
                             "Cancer mutation hotspot", 
                             recurrent_vars_tcga, ~MUTATION_HOTSPOT)
  ),
  list(
    crosstalk::filter_checkbox("LOSS_OF_FUNCTION", "Loss-of-function status", 
                               recurrent_vars_tcga, ~LOSS_OF_FUNCTION),
    crosstalk::filter_slider("TOTAL_RECURRENCE", "Pancancer variant recurrence", 
                             recurrent_vars_tcga, ~TOTAL_RECURRENCE),
    crosstalk::filter_slider("SITE_RECURRENCE", "Site/tissue variant recurrence", 
                             recurrent_vars_tcga, ~SITE_RECURRENCE)
    
  )
)

htmltools::br()
dt <- DT::datatable(recurrent_vars_tcga, escape = F, 
              extensions=c("Buttons","Responsive"), 
              width = "100%",
              options=list(buttons = c('csv','excel'),dom = 'Bfrtip')
)

if(length(mutation_hotspot_levels) > 0){
  dt <- dt %>%
    DT::formatStyle("MUTATION_HOTSPOT","MUTATION_HOTSPOT",
                    color = "white", fontWeight = "bold",
                    backgroundColor = DT::styleEqual(mutation_hotspot_levels, rep("black", length(mutation_hotspot_levels)))
    )
}
dt
htmltools::br()
htmltools::br()


```


```{r tcga_recurrent_variants_missing, echo=F, results = 'asis', eval = NROW(onc_enrich_report[['data']][['tcga']][['recurrent_variants']]) == 0}

cat('\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>NO</b> query proteins with recurrent coding somatic variants in TCGA were found. &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```



#### Copy number alterations

* Genes targeted by somatic copy number alterations (sCNAs) in tumor samples have been retrieved from TCGA, where copy number state have been estimated with [GISTIC](http://portals.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t)
* Gene aberration frequency are plotted in heatmaps across two categories of mutation types
    1. <span style="background-color:#005a32; color:white"><b>&nbsp; sCNA - amplifications &nbsp; </b></span>
    2. <span style="background-color:#800026; color:white"><b>&nbsp; sCNA - homozygous deletions &nbsp; </b></span>
* The values in the heatmaps reflect the percent of all tumor samples pr. primary site with the gene amplified/lost (**percent_mutated**)
  * Genes in the heatmap are ranked according to alteration frequency across all sites (i.e. _pancancer_), limited to the top 75 genes in the query set
* Frequencies across all subtypes per primary site are listed in an interactive table
  * Only including genes that are aberrant in >= 1 percent of samples for a given tumor type/subtype
  * Limited here to top 2,500 gene-subtype frequencies (all aberration frequencies are listed in the Excel output of *oncoEnrichR*)


<br><br>

```{r tcga_ampl_plot, echo = F, eval = !is.null(onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_ampl']])}

suppressMessages(library(plotly))
gene_aberration_top_mat <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_ampl']]
#plotly_colors <- "YlGn"


fig <- plotly::plot_ly(
  colors = "YlGn",
  width = 800,
  height = 400 + (14.67 * NROW(onc_enrich_report$data$tcga$aberration$matrix$cna_ampl))) %>%
    plotly::add_heatmap(
      y = rownames(gene_aberration_top_mat),
      x = colnames(gene_aberration_top_mat),
      z = gene_aberration_top_mat,
      hovertext = "Percent mutated",
      yaxis = "y") %>%
    plotly::layout(
      title = "Amplifications",
      xaxis = list(tickfont = list(size = 13, family = "Helvetica"), 
                   tickangle = -50),
      yaxis = list(tickfont = list(size = 12, family = "Helvetica")),
      margin = list(l = 75, r = 20, b = 150, t = 30, pad = 4),
      legend = list(orientation = 'h')
    ) %>%
    plotly::colorbar(
      nticks = 10,
      title = list(text = "Percent mutated",side = "bottom"),
      limits = c(0, plyr::round_any(max(gene_aberration_top_mat), 10, ceiling)))

rm(gene_aberration_top_mat) 
fig


```

```{r tcga_amplificiations_too_few, echo=F, results = 'asis', eval = is.null(onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_ampl']])}
cat('\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;Heat map not shown - limited number of genes in the queryset are involved in copy number amplifications (<b>< 3, as reported in TCGA cohorts</b>)&nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```

<br><br>

```{r tcga_homdel_plot, echo = F, eval = !is.null(onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_homdel']])}

suppressMessages(library(plotly))
gene_aberration_top_mat <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_homdel']]

#plotly_colors <- "YlOrRd"

fig <- plotly::plot_ly(
  colors = "YlOrRd",
  width = 800,
  height = 400 + (14.67 * NROW(gene_aberration_top_mat))) %>%
    plotly::add_heatmap(
      y = rownames(gene_aberration_top_mat),
      x = colnames(gene_aberration_top_mat),
      z = gene_aberration_top_mat,
      hovertext = "Percent mutated",
      yaxis = "y") %>%
    plotly::layout(
      title = 'Homozygous deletions',
      xaxis = list(tickfont = list(size = 13, family = "Helvetica"), 
                   tickangle = -50),
      yaxis = list(tickfont = list(size = 12, family = "Helvetica")),
      margin = list(l = 75, r = 20, b = 150, t = 30, pad = 4)
    ) %>%
    plotly::colorbar(
      nticks = 10, 
      title = list(text = "Percent mutated",side = "bottom"),
      limits = c(0, plyr::round_any(max(gene_aberration_top_mat), 10, ceiling)))

rm(gene_aberration_top_mat)  
fig


```

```{r tcga_homdel_too_few, echo=F, results = 'asis', eval = is.null(onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_homdel']])}
cat('\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;Heatmap not shown - limited number of genes in the queryset are involved in homozygous deletions (<b>< 3, as reported in TCGA cohorts</b>)&nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


<br><br>




```{r filter_tcga, echo=F, results='asis'}

all_targets_vtypes <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_homdel']] %>% 
  dplyr::bind_rows(
    onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_ampl']]
    ) %>%
  dplyr::arrange(desc(percent_mutated)) %>%
  head(2500)

```


```{r table_browse_tcga_freqs, echo=F, results = "asis", eval = NROW(all_targets_vtypes) > 0}
library(crosstalk)
  
oe_targets_tcga <- crosstalk::SharedData$new(all_targets_vtypes)

crosstalk::bscols(
  list(
    crosstalk::filter_select("primary_site", "Primary site/tissue", oe_targets_tcga, ~primary_site),
    crosstalk::filter_select("variant_type", "Aberration type", oe_targets_tcga, ~variant_type),
    crosstalk::filter_slider("percent_mutated", "Percent mutated", 
                             oe_targets_tcga, ~percent_mutated, step = 10,
                             min = 0, max = 100)
  ),
  list(
    crosstalk::filter_select("primary_diagnosis", "Diagnosis", oe_targets_tcga, ~primary_diagnosis),
    crosstalk::filter_slider("cohort_size", "Cohort Size", oe_targets_tcga, ~cohort_size)
  )
)

htmltools::br()
DT::datatable(oe_targets_tcga, escape = F, 
              extensions=c("Buttons","Responsive"), 
              width = "100%",
              options=list(buttons = c('csv','excel'),dom = 'Bfrtip')
)

```


```{r tcga_freqs_missing, echo=F, results = 'asis', eval = NROW(all_targets_vtypes) == 0}
cat('<i>No genes with tcga aberrations were found.</i>',sep='\n')
cat('\n')
```

<br><br>
