<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Annotation resources • oncoEnrichR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotation resources">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-N4MB4571K5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-N4MB4571K5');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">oncoEnrichR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Intro</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/cancer_gene_rank.html">Cancer gene rank</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/oncoEnrichR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Annotation resources</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/oncoEnrichR/blob/master/vignettes/annotation_resources.Rmd" class="external-link"><code>vignettes/annotation_resources.Rmd</code></a></small>
      <div class="d-none name"><code>annotation_resources.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="resources---overview">Resources - overview<a class="anchor" aria-label="anchor" href="#resources---overview"></a>
</h2>
<p>Data harvested and integrated from the following resources form the
backbone of <em>oncoEnrichR</em>:</p>
<ul>
<li>
<a href="https://targetvalidation.org" class="external-link">Open Targets Platform</a> -
human drug-target associations and comprehensive disease-target
associations</li>
<li>
<a href="http://bionlp.bcgsc.ca/cancermine/" class="external-link">CancerMine</a> -
literature-mined database of drivers, proto-oncogenes and tumor
suppressors in cancer</li>
<li>
<a href="http://ncg.kcl.ac.uk/index.php" class="external-link">Network of Cancer Genes</a>
- Network of cancer genes - a web resource to analyze duplicability,
orthology and network properties of cancer genes</li>
<li>
<a href="https://www.intogen.org" class="external-link">IntOGen</a> - Compendium of
mutational driver genes in cancer</li>
<li>
<a href="https://uniprot.org" class="external-link">UniProt</a> - Comprehensive and freely
accessible resource of protein sequence and functional information</li>
<li><a href="https://cancer.sanger.ac.uk/census" class="external-link">Cancer Gene
Census</a></li>
<li>
<a href="https://saezlab.github.io/dorothea/index.html" class="external-link">DoRothEA</a>
- gene set resource containing signed transcription factor (TF) - target
(regulatory) interactions</li>
<li>
<a href="http://www.cellchat.org/" class="external-link">CellChatDB</a> - ligand-receptor
interaction resource</li>
<li>
<a href="https://portal.gdc.cancer.gov/" class="external-link">The Cancer Genome Atlas
(TCGA)</a> - gene aberration frequencies, recurrent point
mutations/indels, and co-expression patterns in ~9,500 primary tumor
samples</li>
<li>
<a href="https://www.proteinatlas.org" class="external-link">The Human Protein Atlas</a> -
expression data for healthy human tissues (<a href="https://gtexportal.org/home/" class="external-link">GTex</a>)/cell types, and prognostic
gene expression associations in cancer</li>
<li>
<a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp" class="external-link">Molecular
Signatures Database (MSigDB)</a> - collection of annotated (e.g. towards
pathways) genesets for enrichment/over-representation analysis. This
includes genesets from <a href="http://geneontology.org/" class="external-link">Gene
Ontology</a>, <a href="https://reactome.org/" class="external-link">Reactome</a>, <a href="https://www.genome.jp/kegg/pathway.html" class="external-link">KEGG</a>, <a href="https://www.wikipathways.org/index.php/WikiPathways" class="external-link">WikiPathways</a>,
<a href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways" class="external-link">BIOCARTA</a>,
as well as curated <a href="https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C7" class="external-link">immunologic</a>
and <a href="https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C6" class="external-link">cancer-specific</a>
signatures.</li>
<li>
<a href="http://www.netpath.org" class="external-link">NetPath</a> - signaling
transduction pathways</li>
<li>
<a href="https://pfam.xfam.org" class="external-link">Pfam</a> - protein domain
database</li>
<li>
<a href="https://string-db.org" class="external-link">STRING</a> - protein-protein
interaction database</li>
<li>
<a href="https://thebiogrid.org/" class="external-link">BioGRID</a> - biomedical
interaction repository with data compiled through comprehensive curation
efforts</li>
<li>
<a href="https://mips.helmholtz-muenchen.de/corum/" class="external-link">CORUM</a> -
protein complex database</li>
<li>
<a href="https://fgr.hms.harvard.edu/compleat%3E" class="external-link">Compleat</a> -
protein complex resource</li>
<li>
<a href="https://www.ebi.ac.uk/complexportal/home" class="external-link">ComplexPortal</a>
- manually curated, encyclopaedic resource of macro-molecular
complexes</li>
<li>
<a href="http://humap2.proteincomplexes.org" class="external-link">hu.MAP2</a> - human
protein complex map</li>
<li>
<a href="https://compartments.jensenlab.org/Search" class="external-link">COMPARTMENTS</a>
- subcellular compartment database</li>
<li>
<a href="https://score.depmap.sanger.ac.uk" class="external-link">DepMap/Project Score</a>
- Databases on the effects on cancer cell line viability elicited by
CRISPR-Cas9 mediated gene activation</li>
<li>
<a href="https://www.tcga-survival.com/" class="external-link">Genetic determinants of
survival in cancer</a> - Resource on the relative prognostic impact of
gene mutation, expression, methylation, and CNA in human cancers</li>
<li>
<a href="https://pubmed.ncbi.nlm.nih.gov/34529928/" class="external-link">Predicted
synthetic lethality interactions</a> - Resource on predicted synthetic
lethality interactions between human gene paralogs (using human cancer
cell lines)</li>
</ul>
</div>
<div class="section level2">
<h2 id="resources---annotation-types">Resources - annotation types<a class="anchor" aria-label="anchor" href="#resources---annotation-types"></a>
</h2>
<p>This section describes the underlying annotation types of
<em>oncoEnrichR</em> in detail, and to what extent they are subject to
quality control/filtering for use within the package.</p>
<div class="section level3">
<h3 id="target-disease-associations">Target-disease associations<a class="anchor" aria-label="anchor" href="#target-disease-associations"></a>
</h3>
<ul>
<li>Associations between genes and diseases/tumor types are harvested
from the <a href="https://targetvalidation.org" class="external-link">Open Targets
Platform</a>. To increase the confidence of target-disease associations
shown, we include only associations with support from at least two <a href="https://platform-docs.opentargets.org/associations" class="external-link">data
types</a>. Very weak associations, i.e. those with an overall score less
than 0.05, have been excluded (complete range 0-1).</li>
</ul>
</div>
<div class="section level3">
<h3 id="target-drug-associations">Target-drug associations<a class="anchor" aria-label="anchor" href="#target-drug-associations"></a>
</h3>
<ul>
<li>Target-drug associations are primarily harvested from the <a href="https://targetvalidation.org" class="external-link">Open Targets Platform</a>. Through
the use of our own <a href="https://github.com/sigven/phenOncoX" class="external-link">phenOncoX R package</a>, a
dedicated resource on phenotype ontology mapping in cancer, we ensure
that target-drug associations included in <em>oncoEnrichR</em> are
exclusively for cancer conditions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="tumor-suppressorproto-oncogenecancer-driver-annotation">Tumor suppressor/proto-oncogene/cancer driver annotation<a class="anchor" aria-label="anchor" href="#tumor-suppressorproto-oncogenecancer-driver-annotation"></a>
</h3>
<ul>
<li>
<p>We assign three confidence levels with respect to genes having
oncogenic and/or tumor suppressive roles</p>
<ul>
<li>
<strong>Very strong</strong>
<ul>
<li>
<em>Strong_support (&gt; 20 citations)</em> in the biomedical
literature (<a href="http://bionlp.bcgsc.ca/cancermine/" class="external-link">CancerMine</a>), no curated
entry in <a href="http://ncg.kcl.ac.uk/" class="external-link">Network of Cancer Genes
(NCG)</a> or <a href="https://cancer.sanger.ac.uk/census" class="external-link">Cancer Gene
Census (CGC)</a>, <strong>OR</strong>
</li>
<li>Manually curated entry in NCG or CGC, and with <em>significant
support (&gt; 10 citations)</em> in the biomedical literature
(CancerMine)</li>
</ul>
</li>
<li>
<strong>Strong</strong>
<ul>
<li>Curated entry in NCG or CGC, with <em>moderate support (&lt;= 10
citations, &gt; 5 citations)</em> in the biomedical literature
(CancerMine), <strong>OR</strong>
</li>
<li>
<em>Significant support (&gt; 10 citations, &lt;= 20 citations)</em>
in the biomedical literature (CancerMine), no entry in CGC/NCG</li>
</ul>
</li>
<li>
<strong>Moderate</strong>
<ul>
<li>Curated entry in NCG/CGC, with <em>no or weak support (&lt;= 5
citations)</em> in the biomedical literature (CancerMine),
<strong>OR</strong>
</li>
<li>
<em>Moderate support (&lt;= 10 citations, &gt; 5 citations)</em> in
the biomedical literature (CancerMine), no entry in CGC/NCG</li>
</ul>
</li>
<li>Literature support for a tumor suppressive role is ignored if the
gene has 2.5 times as much literature support for an oncogenic role (and
<em>vice versa</em>)</li>
</ul>
</li>
<li>
<p>Classification of genes as potential cancer drivers <em>require
support from at least two of the following sources</em>:</p>
<ul>
<li>
<a href="http://ncg.kcl.ac.uk/" class="external-link">Network of Cancer Genes</a> -
canonical cancer drivers</li>
<li>
<a href="https://www.intogen.org" class="external-link">IntOGen</a> - predicted cancer
driver genes</li>
<li>
<a href="https://cancer.sanger.ac.uk/census" class="external-link">Cancer Gene Census -
TIER1/2</a> - curated cancer genes</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/29625053/" class="external-link">TCGA’s predicted
cancer driver genes</a></li>
<li>
<a href="https://www.nature.com/articles/s41592-019-0422-y" class="external-link">CancerMine</a>
- literature mining resource, considering only entries with support from
&gt; 5 citations)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="gene-cancer-hallmark-associations">Gene cancer hallmark associations<a class="anchor" aria-label="anchor" href="#gene-cancer-hallmark-associations"></a>
</h3>
<ul>
<li>The <a href="https://pubmed.ncbi.nlm.nih.gov/21376230/" class="external-link">hallmarks of
cancer</a> comprise six biological capabilities acquired during the
multistep development of human tumors. Genes attributed to the hallmarks
of cancer have been retrieved from the <a href="https://targetvalidation.org" class="external-link">Open Targets
Platform/COSMIC</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="gene-copy-number-aberrations-in-tumors-somatic-amplificationshomozygous-deletions">Gene copy number aberrations in tumors (somatic
amplifications/homozygous deletions)<a class="anchor" aria-label="anchor" href="#gene-copy-number-aberrations-in-tumors-somatic-amplificationshomozygous-deletions"></a>
</h3>
<ul>
<li>Somatic copy number amplifications and homozygous deletions have
been retrieved from <a href="https://portal.gdc.cancer.gov/" class="external-link">The Cancer
Genome Atlas (TCGA)</a>, focusing on gene-level calls processed with
their <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/CNV_Pipeline/" class="external-link">copy
number variation analysis pipeline</a>. Copy number events included in
<em>oncoEnrichR</em> are limited to high-level amplifications (GISTIC2
CNA value: <strong>2</strong>) and homozygous deletions (GISTIC2 CNA
value: <strong>-2</strong>)</li>
</ul>
</div>
<div class="section level3">
<h3 id="recurrent-gene-mutations-in-tumors-somatic-snvsindels">Recurrent gene mutations in tumors (somatic SNVs/InDels)<a class="anchor" aria-label="anchor" href="#recurrent-gene-mutations-in-tumors-somatic-snvsindels"></a>
</h3>
<ul>
<li>Recurrent somatic SNVs/InDels (occurring in more than one tumor
sample) have been retrieved from <a href="https://portal.gdc.cancer.gov/" class="external-link">The Cancer Genome Atlas
(TCGA)</a>. SNVs/InDels have been annotated further with information on
mutation hotspots in cancer (<a href="https://cancerhotspots.org" class="external-link">cancerhotspots.org</a>), protein
domains from <a href="https://pfam.xfam.org" class="external-link">PFAM</a>, and their
predicted status as loss-of-function variants (<a href="https://github.com/konradjk/loftee" class="external-link">LOFTEE</a>). Only
protein-altering and canonical splice site mutations are included.</li>
</ul>
</div>
<div class="section level3">
<h3 id="gene-co-expression-in-tumors-rna-seq">Gene co-expression in tumors (RNA-seq)<a class="anchor" aria-label="anchor" href="#gene-co-expression-in-tumors-rna-seq"></a>
</h3>
<ul>
<li>Co-expression correlation coefficients between genes in tumors have
been pre-calculated, using RNAseq data available in <a href="https://portal.gdc.cancer.gov/" class="external-link">The Cancer Genome Atlas
(TCGA)</a>. We calculate separate correlation coefficients for each
tumor type, and we further more limit pairs of co-expressed genes to
those that include either a tumor suppressor, an oncogene, or a
predicted cancer driver (using the classification outlined above). Only
strong correlation coefficients (Spearman’s rank &gt;= 0.7 or &lt;=
-0.7) are included</li>
</ul>
</div>
<div class="section level3">
<h3 id="molecular-gene-signatures-pathways">Molecular gene signatures / pathways<a class="anchor" aria-label="anchor" href="#molecular-gene-signatures-pathways"></a>
</h3>
<ul>
<li>A comprehensive set of pathway annotations and molecular gene
signatures form the basis for the functional enrichment module offered
by <em>oncoEnrichR</em>. We include a diverse set of molecular
signatures from <a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link">MSigDB</a>, annotations
with respect to molecular function, biological processes and cellular
components from <a href="http://geneontology.org/" class="external-link">Gene Ontology
(GO)</a>, and pathway annotations from <a href="https://www.genome.jp/kegg/pathway.html" class="external-link">KEGG</a>, <a href="https://www.wikipathways.org/" class="external-link">WikiPathways</a>, <a href="http://netpath.org" class="external-link">NetPath</a>, and <a href="https://reactome.org/" class="external-link">Reactome</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="gene-fitness-effects">Gene fitness effects<a class="anchor" aria-label="anchor" href="#gene-fitness-effects"></a>
</h3>
<ul>
<li>Gene fitness effects are retrieved from an integration of data in <a href="https://score.depmap.sanger.ac.uk/" class="external-link">DepMap &amp; Project
Score</a>. We include data on genes with a statistically significant
effect on cell fitness in cancer cell lines (fitness score is here
considered a quantitative measure of the reduction of cell viability
elicited by a gene inactivation, via CRISPR/Cas9 targeting).</li>
</ul>
</div>
<div class="section level3">
<h3 id="subcellular-compartment-annotation">Subcellular compartment annotation<a class="anchor" aria-label="anchor" href="#subcellular-compartment-annotation"></a>
</h3>
<ul>
<li>Subcellular localization/compartment annotations have been retrieved
from <a href="https://compartments.jensenlab.org/Search" class="external-link">COMPARTMENTS</a>. When
using oncoEnrichR, users have the possibility to configure the
stringency with respect to the target compartment annotations retrieved:
<ul>
<li>Confidence score for each target compartment annotation (range from
3 to 5)</li>
<li>The type of <strong>channels</strong> that support annotations (
<em>Knowledge</em>, <em>Text mining</em> or <em>Experimental</em>), as
well as the required minimum number of channels that should support a
given target compartment annotation</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="protein-complexes-and-protein-domains">Protein complexes and protein domains<a class="anchor" aria-label="anchor" href="#protein-complexes-and-protein-domains"></a>
</h3>
<ul>
<li>Protein complex annotations are retrieved from multiple resources,
including <a href="https://mips.helmholtz-muenchen.de/corum/" class="external-link">CORUM</a>,
<a href="https://fgr.hms.harvard.edu/compleat%3E" class="external-link">Compleat</a>, <a href="https://www.ebi.ac.uk/complexportal/home" class="external-link">ComplexPortal</a>, and
<a href="http://humap2.proteincomplexes.org" class="external-link">hu.MAP2</a>. Protein
domains are retrieved from <a href="https://pfam.xfam.org" class="external-link">Pfam</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="protein-protein-interactions">Protein-protein interactions<a class="anchor" aria-label="anchor" href="#protein-protein-interactions"></a>
</h3>
<ul>
<li>
<p>Protein-protein interactions for members of the target set is
explored using data from two different resources:</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://string-db.org/" class="external-link">STRING</a> - a database of known
and predicted protein-protein interactions, including direct (physical)
and indirect (functional) associations. In oncoEnrichR, the publicly
available STRING <a href="https://string-db.org/cgi/help?subpage=api" class="external-link">API</a> is utilized
for network retrieval, and the users have the possibility to configure
the following:
<ul>
<li>A required <a href="https://string-db.org/cgi/info" class="external-link">minimum
interaction score</a>, limiting the interactions retrieved to those with
an interaction score greater or equal than this threshold. Interaction
scores are indicators of <strong>confidence</strong>, i.e. how likely
STRING judges an interaction to be true.</li>
<li>The <a href="https://string-db.org/cgi/help.pl?subpage=api%23getting-the-string-network-interactions" class="external-link">type
of network</a> to retrieve, either <strong>physical</strong> or
<strong>functional</strong>
</li>
<li>The addition of proteins (maximum 50) that are not part of the query
set, but that interact most confidently with proteins in the query
set</li>
</ul>
</li>
<li>
<a href="https://thebiogrid.org/" class="external-link">BioGRID</a> - a biomedical
interaction repository with data compiled through comprehensive curation
efforts
<ul>
<li>A required <em>minimum number of evidence items</em> that support
each interaction (i.e. from multiple types of low or high-throughput
experiments, e.g. affinity capture-MS, or affinity capture-Western).
Default: 3, maximum: 10</li>
<li>The addition of proteins (maximum 50) that are not part of the query
set, but that interact most confidently with proteins in the query
set</li>
</ul>
</li>
</ol>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="gene-regulatory-interactions">Gene regulatory interactions<a class="anchor" aria-label="anchor" href="#gene-regulatory-interactions"></a>
</h3>
<ul>
<li>Regulatory interactions (TF-target) have been retrieved from the <a href="https://saezlab.github.io/dorothea/" class="external-link">DoRothEA</a> resource. Two
collections of interactions are subject to overlap against the query set
in <em>oncoEnrichR</em>: a <strong>global set</strong> of regulatory
interactions, many of which are inferred from gene expression patterns
in normal tissues (GTEx), and a <strong>cancer-focused set</strong>, the
latter inferred from gene expression patterns in tumors (TCGA).</li>
</ul>
</div>
<div class="section level3">
<h3 id="prognostic-gene-associations">Prognostic gene associations<a class="anchor" aria-label="anchor" href="#prognostic-gene-associations"></a>
</h3>
<ul>
<li>Prognostic associations (gene expression versus survival) are
collected from <a href="https://www.proteinatlas.org/humanproteome/pathology" class="external-link">The Human
Pathology Atlas</a>, which have undertaken correlation analyses of mRNA
expression levels of human genes in tumor tissue and the clinical
outcome (survival) for ~8,000 cancer patients (underlying data from
TCGA). Correlation analyses resulted in more than 10,000 prognostic
genes. We show prognostic genes with regard to cancer type and whether
they are favorable or unfavorable in terms of clinical outcome.
Furthermore, we show prognostic associations established by <a href="https://tcga-survival.com" class="external-link">tcga-survival.com</a>, which relates
various types of gene aberrations (expression or methylation levels,
CNA/mutation status) in tumor samples to patient survival.</li>
</ul>
</div>
<div class="section level3">
<h3 id="ligand-receptor-interactions">Ligand-receptor interactions<a class="anchor" aria-label="anchor" href="#ligand-receptor-interactions"></a>
</h3>
<ul>
<li>Ligand-receptor interactions have been collected from the <a href="http://www.cellchat.org" class="external-link">CellChatDB</a> resource. In the
<em>oncoEnrichR</em> module for these interactions, only interactions
where <strong>both ligand and receptor are found in the query
set</strong> are shown.</li>
</ul>
</div>
<div class="section level3">
<h3 id="synthetic-lethality-interactions">Synthetic lethality interactions<a class="anchor" aria-label="anchor" href="#synthetic-lethality-interactions"></a>
</h3>
<ul>
<li>Predicted synthetic lethality interactions between human gene
paralogs have been collected from <a href="https://pubmed.ncbi.nlm.nih.gov/34529928/" class="external-link">De Kegel et al., Cell
Systems, 2021</a>. Interactions are shown for which both members are
part of the query set, and interactions where only a single member of
the query set is present.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
