<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Changelog • oncoEnrichR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-N4MB4571K5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-N4MB4571K5');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">oncoEnrichR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Intro</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/cancer_gene_rank.html">Cancer gene rank</a></li>
  </ul>
</li>
<li class="active nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/oncoEnrichR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/oncoEnrichR/blob/main/vignettes/CHANGELOG.Rmd" class="external-link"><code>vignettes/CHANGELOG.Rmd</code></a></small>
      <div class="d-none name"><code>CHANGELOG.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="version-1-5-2">Version 1.5.2<a class="anchor" aria-label="anchor" href="#version-1-5-2"></a>
</h2>
<ul>
<li><p>Date: 2024-09-09</p></li>
<li>
<p>Updates</p>
<ul>
<li>Updated pre-processing code that pulls in missing pathway
annotations (KEGG)</li>
<li>Data updates
<ul>
<li>KEGG - 20240904</li>
<li>WikiPathways - 20240810 (fixed bug in report display)</li>
<li>MSigDB (including GO) - v2024.1.Hs</li>
<li>COMPARTMENTS - 2024-09</li>
<li>BIOGRID - v4.4.237</li>
</ul>
</li>
<li>Created a dedicated function <code><a href="../reference/cancer_association_rank.html">cancer_association_rank()</a></code>,
which ranks a query gene set according to strength of association (using
data from the Open Targets Platform) to a primary tumor type/site.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="version-1-5-1">Version 1.5.1<a class="anchor" aria-label="anchor" href="#version-1-5-1"></a>
</h2>
<ul>
<li><p>Date: 2024-08-03</p></li>
<li>
<p>Updates</p>
<ul>
<li>Minor patch that highlights <code>quarto</code> as package
dependency in installation instructions</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="version-1-5-0">Version 1.5.0<a class="anchor" aria-label="anchor" href="#version-1-5-0"></a>
</h2>
<ul>
<li>Date: 2024-08-02</li>
</ul>
<div class="section level3">
<h3 id="changed">Changed<a class="anchor" aria-label="anchor" href="#changed"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>Open Targets Platform (2024.06)</li>
<li>WikiPathways (20240710)</li>
<li>COMPARTMENTS (2024-07)</li>
<li>KEGG (20240626)</li>
<li>BioGRID (v4.4.236)</li>
<li>Upgraded <a href="https://github.com/sigven/phenOncoX" class="external-link">sigven/phenOncoX</a> to
v0.7.6</li>
</ul>
</li>
<li>Functionality
<ul>
<li>Cell/tissue-type enrichment is no longer available (TissueEnrich
package is deprecated)</li>
<li>HTML report is now based on the <a href="https://quarto.org" class="external-link">quarto
framework</a> rather than RMarkdown</li>
<li>Many modules are now disabled by default in order to reduce size of
HTML report output when running in default mode (primarily a Galaxy
consideration)</li>
<li>Limit protein complexes to those harboring at least two members from
the query set</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-4-2">Version 1.4.2<a class="anchor" aria-label="anchor" href="#version-1-4-2"></a>
</h2>
<ul>
<li>Date: 2023-08-16</li>
</ul>
<div class="section level3">
<h3 id="changed-1">Changed<a class="anchor" aria-label="anchor" href="#changed-1"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>WikiPathways (20230810)</li>
<li>COMPARTMENTS (2023-08)</li>
<li>KEGG (20230710)</li>
<li>BioGRID (v4.4.224)</li>
<li>Open Targets Platform (2023.06)</li>
<li>Upgraded <a href="https://github.com/sigven/phenOncoX" class="external-link">sigven/phenOncoX</a> to
v0.6.3:
<ul>
<li>Experimental Factor Ontology (v3.56.0)</li>
<li>Human Disease Ontology (2023-07-20)</li>
</ul>
</li>
<li>A limited set of TCGA amplifications/deletions were erroneously left
out in v1.4.1 (gene alias matching) - fixed in v1.4.2</li>
</ul>
</li>
<li>Functionality
<ul>
<li>Updated <em>Citation Note</em> sections in report</li>
<li>Fixed some typos in the report template</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-4-1">Version 1.4.1<a class="anchor" aria-label="anchor" href="#version-1-4-1"></a>
</h2>
<ul>
<li>Date: 2023-06-18</li>
</ul>
<div class="section level3">
<h3 id="changed-2">Changed<a class="anchor" aria-label="anchor" href="#changed-2"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>TCGA (20220329 - v37.0)</li>
<li>WikiPathways (20230610)</li>
<li>CancerMine (v50, 20230301)</li>
<li>COMPARTMENTS (2023-05)</li>
<li>KEGG (20230330)</li>
<li>BioGRID (v4.4.222)</li>
<li>Open Targets Platform (2023.02)</li>
<li>Cancer Gene Census (v98)</li>
<li>Network of Cancer Genes (v7.1)</li>
<li>MSigDB (v2023.1.Hs)</li>
<li>Upgraded <a href="https://github.com/sigven/phenOncoX" class="external-link">sigven/phenOncoX</a> to
v0.6.1:
<ul>
<li>Experimental Factor Ontology (v3.54.0)</li>
<li>Human Disease Ontology (2023-05-31)</li>
</ul>
</li>
</ul>
</li>
<li>Functionality
<ul>
<li>Added <em>Citation Note</em> section for each output module - key
citations</li>
<li>Added data licensing terms in <em>Documentation &amp; Settings</em>
module</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-4-0">Version 1.4.0<a class="anchor" aria-label="anchor" href="#version-1-4-0"></a>
</h2>
<ul>
<li>Date: 2023-02-21</li>
</ul>
<div class="section level3">
<h3 id="changed-3">Changed<a class="anchor" aria-label="anchor" href="#changed-3"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>TCGA (20221212 - v36.0)</li>
<li>WikiPathways (20230210)</li>
<li>CancerMine (20230114)</li>
<li>COMPARTMENTS (2023-01)</li>
<li>KEGG (20230101)</li>
<li>Upgraded <a href="https://github.com/sigven/phenOncoX" class="external-link">sigven/phenOncoX</a> to
v0.5.6:
<ul>
<li>Experimental Factor Ontology (v3.50.0)</li>
<li>Human Disease Ontology (2023-01-31)</li>
</ul>
</li>
</ul>
</li>
<li>Functionality
<ul>
<li>
<em>Loading of databases</em>
<ul>
<li>Downloading through Google Drive can somethis cause trouble due to
downloading limitations issued by Google. As an alternative approach,
one can now download datasets using another host (server hosted by the
University of Oslo), through the argument
<code>googledrive = FALSE</code> in
<code><a href="../reference/load_db.html">oncoEnrichR::load_db()</a></code>
</li>
</ul>
</li>
<li>
<em>Protein-protein interactions</em>
<ul>
<li>Users can opt to show isolated nodes (nodes with no interactions) or
not - logical argument to <code><a href="../reference/onco_enrich.html">onco_enrich()</a></code>:
<code>ppi_show_isolated_nodes</code> (default <em>FALSE</em>)</li>
<li>Added protein-protein interaction network based on interactions
found in <a href="https://thebiogrid.org" class="external-link">BioGRID</a>
<ul>
<li>Users can set a minimum required threshold for the number of
evidence items required for each interaction in BioGRID</li>
<li>New numerical argument to <code><a href="../reference/onco_enrich.html">onco_enrich()</a></code>:
<code>ppi_biogrid_min_evidence</code>
</li>
</ul>
</li>
<li>Users can restrict the STRING network to <em>physical</em>
interactions only or an extended network also including indirect (
<em>functional</em> ) interactions - <a href="https://string-db.org/help/api/#getting-the-string-network-interactions" class="external-link">API
network type</a>
<ul>
<li>New character argument to <code><a href="../reference/onco_enrich.html">onco_enrich()</a></code>:
<code>ppi_string_network_type</code> (<em>physical</em> or
<em>functional</em>)</li>
</ul>
</li>
<li>Previous argument <code>ppi_score_threshold</code> is now named
<code>ppi_string_min_score</code>, with a scale that goes from 0 to 1
(as opposed to 0-1000 in previous versions)</li>
</ul>
</li>
<li>
<em>Subcellular compartment annotations</em>
<ul>
<li>Now using the regularly updated <a href="https://compartments.jensenlab.org/Search" class="external-link">COMPARTMENTS</a>
resource as the underlying source for annotation</li>
<li>Users can set a minimum confidence score for each target compartment
annotation (range from 3 (least confident) to 5 (most confident)), as
well as the required minimum number of channels ( <em>Knowledge</em>,
<em>Text mining</em> or <em>Experimental</em>) that should support a
given target compartment annotation
<ul>
<li>New arguments to <code><a href="../reference/onco_enrich.html">onco_enrich()</a></code>:
<code>subcellcomp_min_confidence</code>,
<code>subcellcomp_min_channels</code>
</li>
</ul>
</li>
</ul>
</li>
<li>
<em>Annotation of oncogenic/tumor suppressive roles</em>
<ul>
<li>A revised approach to assign roles as proto-oncogenes or tumor
suppressors is implemented, in which we assign each role with a
confidence level (<em>Moderate/Strong/Very strong</em>) pending upon
support in manually curated resources (Network of Cancer Genes (NCG) or
Cancer Gene Census (CGC)) or support in the biomedical literature
(CancerMine)</li>
</ul>
</li>
<li>
<em>Cancer driver annotation</em>
<ul>
<li>A more conservative approach for driver gene status is implemented:
requiring support from <em>at least two</em> distinct resources (Network
of Cancer Genes (NCG), Cancer Gene Census, TCGA, IntOGen, or
CancerMine)</li>
</ul>
</li>
<li>
<em>Argument renaming</em> - Renamed arguments to
<code><a href="../reference/onco_enrich.html">onco_enrich()</a></code> for improved naming consistency:
<ul>
<li><code>ppi_string_min_score</code></li>
<li><code>ppi_string_network_type</code></li>
<li><code>ppi_biogrid_min_evidence</code></li>
<li><code>ppi_show_drugs</code></li>
<li><code>ppi_show_isolated_nodes</code></li>
<li><code>regulatory_min_confidence</code></li>
<li><code>enrichment_p_value_cutoff</code></li>
<li><code>enrichment_q_value_cutoff</code></li>
<li><code>enrichment_p_value_adj</code></li>
<li><code>enrichment_plot_num_terms</code></li>
<li><code>enrichment_simplify_go</code></li>
<li><code>enrichment_max_geneset_size</code></li>
<li><code>enrichment_min_geneset_size</code></li>
<li><code>subcellcomp_min_confidence</code></li>
<li><code>subcellcomp_show_cytosol</code></li>
<li><code>subcellcomp_min_channels</code></li>
</ul>
</li>
<li>
<em>Other</em>
<ul>
<li>added oncoEnrichR favicon to output HTML reports</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-3-2">Version 1.3.2<a class="anchor" aria-label="anchor" href="#version-1-3-2"></a>
</h2>
<ul>
<li>Date: 2022-09-27</li>
</ul>
<div class="section level3">
<h3 id="changed-4">Changed<a class="anchor" aria-label="anchor" href="#changed-4"></a>
</h3>
<ul>
<li>Cancer driver classification in <code>Cancer associations</code>
section. Driver classification is based on the union of IntOGen
mutational driver catalogue (v2020-02-01) and Network of Cancer Genes
(canonical drivers, v7.0)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-3-1">Version 1.3.1<a class="anchor" aria-label="anchor" href="#version-1-3-1"></a>
</h2>
<ul>
<li>Date: 2022-09-23</li>
</ul>
<div class="section level3">
<h3 id="added">Added<a class="anchor" aria-label="anchor" href="#added"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>CancerMine (20220920 - v48)</li>
<li>TCGA (20220727 - v34.0)</li>
<li>WikiPathways (20220910)</li>
<li>Upgraded <a href="https://github.com/sigven/oncoPhenoMap" class="external-link">sigven/oncoPhenoMap</a> to
v0.4.0:
<ul>
<li>Experimental Factor Ontology (v3.46.0)</li>
<li>Human Disease Ontology (2022-08-29)</li>
</ul>
</li>
</ul>
</li>
<li>Recurrent somatic variants (SNVs/InDels, as found in TCGA) are
appended to Excel output worksheet, tab
<code>RECURRENT_VARIANTS</code>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="fixed">Fixed<a class="anchor" aria-label="anchor" href="#fixed"></a>
</h3>
<ul>
<li>A few erroneous mutation hotspots in
<code>Tumor aberration frequencies</code> section</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-5">Changed<a class="anchor" aria-label="anchor" href="#changed-5"></a>
</h3>
<ul>
<li>Slight modification to column names in
<code>Drug associations</code> section</li>
<li>Added links to <code>ENSEMBL_GENE_ID</code> in
<code>Tumor aberration frequencies</code> section</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-3-0">Version 1.3.0<a class="anchor" aria-label="anchor" href="#version-1-3-0"></a>
</h2>
<ul>
<li>Date: 2022-09-12</li>
</ul>
<div class="section level3">
<h3 id="added-1">Added<a class="anchor" aria-label="anchor" href="#added-1"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>MSigDB (August 2022.1)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-6">Changed<a class="anchor" aria-label="anchor" href="#changed-6"></a>
</h3>
<ul>
<li>Fixed some artefacts (<code>-AS1</code> entries) in the output of
CancerMine</li>
<li>Added <code>approved_drugs</code> as a column to the
<code>Drug associations</code> section of the output report</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-2-2">Version 1.2.2<a class="anchor" aria-label="anchor" href="#version-1-2-2"></a>
</h2>
<ul>
<li>Date: 2022-09-02</li>
</ul>
<div class="section level3">
<h3 id="added-2">Added<a class="anchor" aria-label="anchor" href="#added-2"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>WikiPathways (20220810)</li>
<li>KEGG (20220809)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-7">Changed<a class="anchor" aria-label="anchor" href="#changed-7"></a>
</h3>
<ul>
<li>Now using <a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a> as data
repository host, due to unstable code using Zenodo</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-2-1">Version 1.2.1<a class="anchor" aria-label="anchor" href="#version-1-2-1"></a>
</h2>
<ul>
<li>Date: 2022-07-13</li>
</ul>
<div class="section level3">
<h3 id="added-3">Added<a class="anchor" aria-label="anchor" href="#added-3"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>WikiPathways (20220610)</li>
<li>Open Targets Platform (2022.06)</li>
<li>CancerMine (v47)</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-2-0">Version 1.2.0<a class="anchor" aria-label="anchor" href="#version-1-2-0"></a>
</h2>
<ul>
<li>Date: 2022-06-24</li>
</ul>
<div class="section level3">
<h3 id="changed-8">Changed<a class="anchor" aria-label="anchor" href="#changed-8"></a>
</h3>
<ul>
<li>Moved to <a href="https://doi.org/10.5281/zenodo.6684347" class="external-link">zenodo.org</a> for hosting
underlying datasets, accessed with <a href="https://github.com/eblondel/zen4R/" class="external-link">zen4R</a>
</li>
<li>MAF objects (TCGA) no longer loaded from GitHub</li>
<li>Function <code><a href="../reference/write.html">oncoEnrichR::write()</a></code> now requires the
oncoEnrichR database object (<code>oeDB</code>) as an argument</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-1-1">Version 1.1.1<a class="anchor" aria-label="anchor" href="#version-1-1-1"></a>
</h2>
<ul>
<li>Date: 2022-06-16</li>
</ul>
<div class="section level3">
<h3 id="added-4">Added<a class="anchor" aria-label="anchor" href="#added-4"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>WikiPathways (20220610)</li>
</ul>
</li>
<li>Configurations for <code><a href="../reference/onco_enrich.html">onco_enrich()</a></code> populated at the end
of HTML report</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-9">Changed<a class="anchor" aria-label="anchor" href="#changed-9"></a>
</h3>
<ul>
<li>Option renaming
<ul>
<li>
<code>show_prognostic_cancer_assoc</code> renamed to
<code>show_prognostic</code>.</li>
<li>
<code>show_tcga_aberration</code> renamed to
<code>show_aberration</code>
</li>
<li>
<code>show_regulatory_interactions</code> renamed to
<code>show_regulatory</code>
</li>
<li>
<code>show_tcga_coexpression</code> renamed to
<code>show_coexpression</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="fixed-1">Fixed<a class="anchor" aria-label="anchor" href="#fixed-1"></a>
</h3>
<ul>
<li>Bug: database loading missing <code>pfamdb</code>
</li>
<li>Relaxed dependency versions (caused conflicts for R versions &lt;
4.2)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-1-0">Version 1.1.0<a class="anchor" aria-label="anchor" href="#version-1-1-0"></a>
</h2>
<ul>
<li>Date: 2022-06-10</li>
</ul>
<div class="section level3">
<h3 id="added-5">Added<a class="anchor" aria-label="anchor" href="#added-5"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>CancerMine (v46)</li>
<li>Open Targets Platform (2022.04)</li>
<li>WikiPathways (20220510)</li>
<li>GENCODE (v40)</li>
<li>CGC (v96)</li>
<li>TCGA (GDC release 32)</li>
<li>EFO (v42.0)</li>
</ul>
</li>
<li>Module that shows the occurrence of protein domains among query set
members</li>
<li>Protein-complex cancer relevance score</li>
<li>New options
<ul>
<li>
<code>ppi_node_shadow</code> - logical indicating if nodes in the
PPI network should carry a shadow or not</li>
<li>
<code>show_domain</code> - logical indicating if report should add
section on protein domain frequencies in query set</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-10">Changed<a class="anchor" aria-label="anchor" href="#changed-10"></a>
</h3>
<ul>
<li>Re-ordering of HTML sections, multiple cosmetic changes</li>
<li>GO enrichment plots: now showing q-value and enrichment factors in
bar plots</li>
<li>Revised method to compute tumor-type specific cancer rank (and
global)</li>
<li>Dedicated <em>output</em> article with output views on
sigven.github.io/oncoEnrichR</li>
</ul>
</div>
<div class="section level3">
<h3 id="fixed-2">Fixed<a class="anchor" aria-label="anchor" href="#fixed-2"></a>
</h3>
<ul>
<li>Bug: duplicate records in retrieved co-expression pairs</li>
<li>Removed most redundant protein complexes in OmniPath</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-0-9">Version 1.0.9<a class="anchor" aria-label="anchor" href="#version-1-0-9"></a>
</h2>
<ul>
<li>Date: 2022-03-31</li>
</ul>
<div class="section level3">
<h3 id="added-6">Added<a class="anchor" aria-label="anchor" href="#added-6"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>CancerMine (v43)</li>
<li>Open Targets Platform (2022.02)</li>
<li>WikiPathways (20220310)</li>
<li>Project Score (July 2021 release)</li>
</ul>
</li>
<li>New analysis section: <em>Synthetic lethality</em> - shows how
members of the queryset overlaps with predicted synthetic lethality
interactions (as published by De Kegel et al., Cell Systems, 2021)
<ul>
<li>add this section in the output with option
<code>--show_synleth</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="fixed-3">Fixed<a class="anchor" aria-label="anchor" href="#fixed-3"></a>
</h3>
<ul>
<li>Bug in query identification for background set</li>
<li>Bug in rank of top gene-cancer associations</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-11">Changed<a class="anchor" aria-label="anchor" href="#changed-11"></a>
</h3>
<ul>
<li>Option <code>--show_crispr_lof</code> renamed to
<code>--show_fitness</code>. Corresponding section in HTML report
renamed from <code>CRISPR/Cas9 loss-of-fitness</code> to
<code>Gene fitness scores</code>.</li>
<li>Renamed sections in HTML report:
<ul>
<li>
<code>TCGA co-expression</code> –&gt;
<code>Tumor co-expression</code>
</li>
<li>
<code>TCGA aberration frequency</code> —&gt;
<code>Tumor aberration frequencies</code>
</li>
<li>
<code>TCGA prognostic associations</code> —&gt;
<code>Prognostic associations</code>
</li>
</ul>
</li>
<li>Sections that include <em>tabsets</em> are now organized so that the
initial active tab always contains data (with the exception of all tabs
being empty)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-0-8">Version 1.0.8<a class="anchor" aria-label="anchor" href="#version-1-0-8"></a>
</h2>
<ul>
<li>Date: 2022-02-18</li>
</ul>
<div class="section level3">
<h3 id="added-7">Added<a class="anchor" aria-label="anchor" href="#added-7"></a>
</h3>
<ul>
<li>Data updates:
<ul>
<li>WikiPathways (20220210)</li>
<li>CancerMine (v42)</li>
<li>MSigDB (v7.5.1)</li>
<li>GENCODE (v39)</li>
<li>Gene summary (NCBI)</li>
<li>Reactome/GO (MSigDB v7.5.1)</li>
<li>KEGG (20211223)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-12">Changed<a class="anchor" aria-label="anchor" href="#changed-12"></a>
</h3>
<ul>
<li>Refactoring of code and data for mapping gene identifiers</li>
<li>Refactoring of code for database loading</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-0-7">Version 1.0.7<a class="anchor" aria-label="anchor" href="#version-1-0-7"></a>
</h2>
<ul>
<li>Date: 2021-11-30</li>
</ul>
<div class="section level3">
<h3 id="added-8">Added<a class="anchor" aria-label="anchor" href="#added-8"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>WikiPathways (20211110)</li>
<li>Open Targets Platform (2021.11)</li>
<li>EFO (v3.36.0)</li>
<li>TCGA (v31, 2021-10-29)</li>
<li>CancerMine (v40)</li>
<li>Human Protein Atlas (v21)</li>
<li>UniProt KB (2021_04)</li>
<li>PFAM domains (release 35, November 2021)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="changed-13">Changed<a class="anchor" aria-label="anchor" href="#changed-13"></a>
</h3>
<ul>
<li>Management of annotation databases have been re-designed, reducing
the size and installation of <em>oncoEnrichR</em> significantly. The
re-design results in the addition of an initial step in the analysis;
<strong>database loading</strong>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="version-1-0-6">Version 1.0.6<a class="anchor" aria-label="anchor" href="#version-1-0-6"></a>
</h2>
<ul>
<li>Date: 2021-10-27</li>
</ul>
<div class="section level3">
<h3 id="added-9">Added<a class="anchor" aria-label="anchor" href="#added-9"></a>
</h3>
<ul>
<li>Data updates
<ul>
<li>KEGG (20211013)</li>
<li>WikiPathways (20211010)</li>
<li>Open Targets Platform (2021.09)</li>
<li>EFO (v3.35.0), DiseaseOntology (v2021-10-11)</li>
<li>TCGA (v30, 2021-09-23)</li>
<li>Protein complex annotations - additions from <a href="https://fgr.hms.harvard.edu/compleat" class="external-link">Compleat</a>, <a href="https://www.ebi.ac.uk/complexportal/home" class="external-link">ComplexPortal</a>, <a href="https://www.rcsb.org/" class="external-link">PDB</a>, <a href="http://humap2.proteincomplexes.org/" class="external-link">hu.MAP2</a>
</li>
</ul>
</li>
<li>New functionality
<ul>
<li>Report modules
<ul>
<li>
<strong>Regulatory interactions</strong> - (<a href="https://saezlab.github.io/dorothea/" class="external-link">DoRothEA</a>)
<ul>
<li>View overlap between members of queryset with previously established
regulatory interactions (TF-target relationships) from the DoRothEA
resource</li>
</ul>
</li>
<li>
<strong>Ligand-receptor interactions</strong> - (<a href="http://www.cellchat.org/" class="external-link">CellChatDB</a>)
<ul>
<li>Explore potential ligand-receptor interactions in the query set, as
found in the CellChat database</li>
</ul>
</li>
</ul>
</li>
<li>Report styling options
<ul>
<li>Choose your own Bootswatch theme for HTML report - option
<code>html_report_theme</code>
</li>
<li>Choose where the table of contents are placed in the HTML report
(floating-left or static at the top) - option
<code>html_floating_toc</code>
</li>
</ul>
</li>
<li>Arguments to <code>oncoEnichR::onco_enrich()</code>:
<ul>
<li>
<code>html_floating_toc</code> - logical, if set to FALSE, table of
contents in HTML report will be placed on top of the main document</li>
<li>
<code>html_report_theme</code> - character, choose between different
Bootswatch themes for style in HTML report</li>
<li>
<code>show_regulatory</code> - logical, show regulatory interactions
module (DoRothEA)</li>
<li>
<code>min_confidence_reg_interaction</code> - character, minimum
confidence of regulatory interactions included from DoRothEA
(‘A’,‘B’,‘C’, or ‘D’)</li>
<li>
<code>show_ligand_receptor</code> - logical, show ligand-receptor
interaction module (CellChatDB)</li>
<li>
<code>num_terms_enrichment_plot</code> - integer, number of enriched
Gene Ontology terms (max) to show in enrichment barplots (module
<em>Functional Enrichment</em>)</li>
</ul>
</li>
<li>Default gene rankings
<ul>
<li>In the modules <strong>Regulatory Interactions</strong> and
<strong>Tissue and cell type enrichment</strong>, interactions/targets
are ranked according to a quantitative cancer association score
<em>cancer_max_rank</em> (maximum gene-cancer association rank (Open
Targets Platform) across primary sites)</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="fixed-4">Fixed<a class="anchor" aria-label="anchor" href="#fixed-4"></a>
</h3>
<ul>
<li>Report <em>disclaimer</em> occurring at the bottom of the report is
no longer missing from web-based analysis (oncotools.elixir.no) - file
<code>_site.yml</code> in <code>inst/templates</code>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
