<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running • oncoEnrichR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-N4MB4571K5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-N4MB4571K5');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">oncoEnrichR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Intro</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/running.html">Running</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/input.html">Input</a></li>
    <li><a class="dropdown-item" href="../articles/output.html">Output</a></li>
    <li><a class="dropdown-item" href="../articles/annotation_resources.html">Annotation resources</a></li>
    <li><a class="dropdown-item" href="../articles/cancer_gene_rank.html">Cancer gene rank</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">CHANGELOG</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/oncoEnrichR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Running</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sigven/oncoEnrichR/blob/main/vignettes/running.Rmd" class="external-link"><code>vignettes/running.Rmd</code></a></small>
      <div class="d-none name"><code>running.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="key-functions">Key functions<a class="anchor" aria-label="anchor" href="#key-functions"></a>
</h2>
<p><em>oncoEnrichR</em> performs its operations through the following
three functions:</p>
<div class="section level3">
<h3 id="oncoenrichrload_db">1. <code>oncoEnrichR::load_db()</code><a class="anchor" aria-label="anchor" href="#oncoenrichrload_db"></a>
</h3>
<ul>
<li>Loads the underlying annotation data repository for oncoEnrichR, and
saves it to a local cache directory.</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="oncoenrichronco_enrich">2. <code>oncoEnrichR::onco_enrich()</code><a class="anchor" aria-label="anchor" href="#oncoenrichronco_enrich"></a>
</h3>
<ul>
<li>
<p>Consists of two main processing steps:</p>
<p>1) Takes an input/query list of human gene/protein identifiers
(e.g. UniProt accession, RefSeq/Ensembl transcript identifer etc.) as
input and conducts uniform identifier conversion</p>
<p>2) Performs extensive annotation, enrichment and membership analyses
of the query set against underlying data sources on cancer-relevant
properties of human genes and their interrelationships.</p>
</li>
<li>
<p>Technically, the method returns a <em>list object</em> with all
contents of the analyses performed. The specific arguments/options and
default values are outlined below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/onco_enrich.html">onco_enrich</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  oeDB <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query_id_type <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  ignore_id_err <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  project_title <span class="op">=</span> <span class="st">"_Project title_"</span>,</span>
<span>  project_owner <span class="op">=</span> <span class="st">"_Project owner_"</span>,</span>
<span>  project_description <span class="op">=</span> <span class="st">"_Project description_"</span>,</span>
<span>  bgset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bgset_id_type <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  bgset_description <span class="op">=</span> <span class="st">"All protein-coding genes"</span>,</span>
<span>  enrichment_p_value_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  enrichment_p_value_adj <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  enrichment_q_value_cutoff <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  enrichment_min_geneset_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  enrichment_max_geneset_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  enrichment_plot_num_terms <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  enrichment_simplify_go <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  subcellcomp_min_confidence <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  subcellcomp_min_channels <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  subcellcomp_show_cytosol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  regulatory_min_confidence <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>  fitness_max_score <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  ppi_add_nodes <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  ppi_string_min_score <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  ppi_string_network_type <span class="op">=</span> <span class="st">"functional"</span>,</span>
<span>  ppi_biogrid_min_evidence <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  ppi_node_shadow <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ppi_show_drugs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ppi_show_isolated_nodes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_ppi <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_disease <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_top_diseases_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_cancer_hallmarks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_drug <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_enrichment <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_aberration <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_coexpression <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_ligand_receptor <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_regulatory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_unknown_function <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_prognostic <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_subcell_comp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_synleth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_fitness <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_complex <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_domain <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
<p>See <a href="https://sigven.github.io/oncoEnrichR/reference/onco_enrich.html">detailed
descriptions of all options here</a> Note that not all modules of the
report are enabled by default. The user can enable or disable specific
modules by setting the corresponding <code>show_</code> arguments to
<code>TRUE</code> or <code>FALSE</code>.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="oncoenrichrwrite">3. <code>oncoEnrichR::write()</code><a class="anchor" aria-label="anchor" href="#oncoenrichrwrite"></a>
</h3>
<ul>
<li>
<p>Consists of two main processing steps:</p>
<p>1) Transformation of the raw analysis results returned by
<em>oncoEnrichR::onco_enrich()</em> into various visualizations and
interactive tables</p>
<p>2) Assembly and generation of the final analysis report through</p>
<ul>
<li>A) a structured and interactive <em>oncoEnrichR</em> HTML
report</li>
<li>B) a multisheet Excel workbook</li>
</ul>
</li>
</ul>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="example-run">Example run<a class="anchor" aria-label="anchor" href="#example-run"></a>
</h2>
<p>A target list of <em>n = 134</em> high-confidence interacting
proteins with the c-MYC oncoprotein were previously identified through
BioID protein proximity assay in standard cell culture and in tumor
xenografts (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25452129" class="external-link">Dingar et al., J
Proteomics, 2015</a>). We ran this target list through the
<em>oncoEnrichR</em> analysis workflow using the following
configurations for the <code>onco_enrich</code> method:</p>
<ul>
<li><code>project_title = "cMYC_BioID_screen"</code></li>
<li><code>project_owner = "Raught et al."</code></li>
</ul>
<p>and produced the <a href="https://doi.org/10.5281/zenodo.13734871" class="external-link">following HTML report
with results</a>.</p>
<p>Below are R commands provided to reproduce the example output.
<strong>NOTE</strong>: Replace “&lt;LOCAL_FOLDER&gt;” with a directory
on your local computer:</p>
<ul>
<li><code><a href="https://sigven.github.io/oncoEnrichR">library(oncoEnrichR)</a></code></li>
<li><code>myc_interact_targets &lt;- read.csv(system.file("extdata","myc_data.csv", package = "oncoEnrichR"), stringsAsFactors = F)</code></li>
<li><code>oeDB &lt;- oncoEnrichR::load_db(cache_dir = "&lt;LOCAL_FOLDER&gt;")</code></li>
<li><code>myc_report &lt;- oncoEnrichR::onco_enrich(query = myc_interact_targets$symbol, oeDB = oeDB, project_title = "cMYC BioID screen", project_owner = "Raught et al.")</code></li>
<li><code>oncoEnrichR::write(report = myc_report, oeDB = oeDB, file = "&lt;LOCAL_FOLDER&gt;/myc_report_oncoenrichr.html", format = "html")</code></li>
<li><code>oncoEnrichR::write(report = myc_report, oeDB = oeDB, file = "&lt;LOCAL_FOLDER&gt;/myc_report_oncoenrichr.xlsx", format = "excel")</code></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/sigven" class="external-link">Sigve Nakken</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
